### ISSUE TITLE
Layout commands aren't aware of shell aliases / sirens

### ISSUE THREAD

**Issue by yoav-lavi** [zellij-org/zellij#2705]
I use Zellij under Nix and have run into a few issues with layout pane commands:

1. Setting up a shell function (in Fish) does not make that function available to `command="xyz"`
2. Commands run by `command="xyz"` do not use a local nix flake via nix-direnv, even if Nix is set up to do so (which means e.g. running your editor via `command` runs an editor without the local configured language (e.g. Rust))

This is probably related to https://github.com/zellij-org/zellij/issues/2653

Thanks!

**Comment by scabala** [comment]
That is no a bug, that is a feature.

What you can set as `command` is a system command, namely any executable found in `PATH`. In contrast to other tools, zellij does not run commands through shell.

If you want to use shell, your best bet is to use `command=sh -c "xyz"`

**Comment by Zykino** [comment]
Yes, exactly. By default zellij start the commands/program itself (not using a shell).

I think the workaround is `pane command=fish { args "-c" "xyz" }` (just to be a bit more precise, using `args` and the example shell given in the opening of the issue).

**Comment by yoav-lavi** [comment]
I see, thanks!

Anecdotally this didn't work when using args="abc def ghi" but from your comments I understand that these work like Rust's `Command::new(...).args` / `.arg()` and can't have spaces in a single item (perhaps specifically because they're implemented using that or a similar API, or just for the same security considerations)

Will try this again soon with that in mind

**Comment by yoav-lavi** [comment]
Update: this does not seem to work either

**Comment by Zykino** [comment]
I’m a bit confused. What did not work? The `args="abc def ghi"` are interpreted as 1 argument?
Also as said, for your command to be recognized it should be in the `PATH` (of the shell that start the zellij’s server).

**Comment by yoav-lavi** [comment]
Sorry, allow me to clarify:

Using something like

```idl
pane {
  command "fish"
  args "hx" "."
}
```

Still does not seem to load direnv properly.

Theoretically since home-manager is managing Fish (by modifying its config file) it should load when run in the same directory via nix-direnv but for some reason when Zellij is the one running the shell it doesn't.

**Comment by yoav-lavi** [comment]
Edit: fixed the sample above, made it from memory and forgot a flag

**Comment by yoav-lavi** [comment]
Also, 

```kdl
pane {
  command "fish"
  args "-c" "piman" // persistant - interactive - manual, basically a while true loop of apropos . -> fzf -> man
}
```

Did not seem to properly use the alias I have set

**Comment by Zykino** [comment]
Hmmm I re-read your original issue. I think you need those (from `fish`’s man page):

> -c or --command=COMMANDS evaluate the specified commands instead of reading from the commandline, passing any additional positional arguments via $argv. Note that, unlike other shells, the first argument is not the name of the program ($0), but simply the first normal argument.
> -C or --init-command=COMMANDS evaluate the specified commands after reading the configuration, before running the command specified by -c or reading interactive input

So first the easy : you forgot `-c` before the command (small C not capital one).

Then I don’t really know direnv nor nix (and each time I forget what peoples find as answer). But I suspect the `-C` option (capital C this time) can help you start direnv before your command. Also, it may not be needed but your direnv autoload need to be outside the "interactive only" portion of fish’s config.

Lastly, if you want to start helix, you can also use `pane edit="file-path.txt"` instead of starting a command (or starting a shell that start your editor).

**Comment by Zykino** [comment]
Ain’t alias interactive only in fish?

**Comment by yoav-lavi** [comment]
@Zykino I edited the sample right after posting it, I just typed it from memory and forgot the "-c" but it was there originally 😄

**Comment by yoav-lavi** [comment]
>Ain’t alias interactive only in fish?
WDYM?

**Comment by Zykino** [comment]
Create a simple shell script and put it in `~/.local/bin`. Add this to your $PATH ^^

**Comment by yoav-lavi** [comment]
>Create a simple shell script and put it in ~/.local/bin. Add this to your $PATH ^^

Not sure I follow, what is this in response to?

**Comment by yoav-lavi** [comment]
>pane edit="file-path.txt"

The issue here is that I need Helix to also use direnv, since otherwise I won't have access to rust-analyzer / the tooling for my projects (which are nix flakes in the project direcotry)

**Comment by yoav-lavi** [comment]
I'll try `-C` and running the shell as interactive, perhaps it's just skipping a bit of configuration that's required

**Comment by yoav-lavi** [comment]
What's the reasoning for not allowing a command to be run in the current shell in a pane? Perhaps that can use a different config?

It also stands to reason that people might want an initial command but not for it to reload / not allow exiting to shell

**Comment by yoav-lavi** [comment]
(to be clear I'm not suggesting the decision was wrong, just justifying it as a possible addition)

**Comment by yoav-lavi** [comment]
Update: `-i` doesn't work for direnv but does for aliases.

I'll try `-C` now to manually use direnv

**Comment by yoav-lavi** [comment]
That doesn't seem to work either

**Comment by yoav-lavi** [comment]
That doesn't seem to work either, both with or without `-i`

**Comment by Zykino** [comment]
I mixed fish’s alias (which are functions) and abbreviation https://fishshell.com/docs/current/cmds/abbr.html
> **Note:** Only typed-in commands use abbreviations. Abbreviations are not expanded in scripts. 

In a shell the interactive mode is the standard one (when you start it only with its name `bash`, `fish`). Here the `-i` force this standard mode.

In fish you can easily differentiate between your interactive (standard) shell and your command (script) one:
```fish
❯ cat ~/.config/fish/config.fish
if status is-interactive
# …
end
```

I think instead of your alias you are better of creating a script file. Putting it in the PATH (before starting `zellij``s session) and maybe your alias call this script.
Then if you start `bash` and achieve what you want with `fish -c …` you just have to past that into your layout file.

Also, someone on discord recently fought with direnv in nix too, but to start a command/pane. Apparently their workaround was `zellij run -- direnv exec . <binary>`

So instead of trying to run fish maybe use:
```kdl
pane command="direnv" {
    args "exec" "." "hx"
}
```

Not sure if you need to keep the "." or re-past your directory.

**Comment by ck3mp3r** [comment]
This is not only an issue with fish. I've stumbled across this myself. Layout command executions do not seem to load the environment before executing. I have a project setup with direnv. A devshell is started once entering the directory and tooling becomes available from that point onwards. A simple command like "zola serve" just doesn't work because zola is not available on the global path...
